File: src/main/java/io/github/romeoahmed/cursedoath/domain/BarrierState.java
package io.github.romeoahmed.cursedoath.domain;

import io.github.romeoahmed.cursedoath.CursedOath;
import net.fabricmc.fabric.api.attachment.v1.AttachmentRegistry;
import net.fabricmc.fabric.api.attachment.v1.AttachmentSyncPredicate;
import net.fabricmc.fabric.api.attachment.v1.AttachmentType;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.world.entity.Entity;

/// Synced, unsaved defense state: `0` is inactive, positive values are Simple Domain strength,
/// and `-1` is Domain Amplification.
public final class BarrierState {
    private BarrierState() {}

    private static final AttachmentType<Integer> STATE = AttachmentRegistry.create(
            CursedOath.id("barrier_state"),
            builder -> builder.initializer(() -> 0).syncWith(ByteBufCodecs.VAR_INT, AttachmentSyncPredicate.all()));

    public static void initialize() {}

    public static int get(Entity entity) {
        var state = entity.getAttached(STATE);
        return state == null ? 0 : state;
    }

    public static void update(Entity entity, int value) {
        if (get(entity) != value) entity.setAttached(STATE, value);
    }
}

File: src/main/kotlin/io/github/romeoahmed/cursedoath/domain/BarrierState.kt
package io.github.romeoahmed.cursedoath.domain

import io.github.romeoahmed.cursedoath.CursedOath
import net.fabricmc.fabric.api.attachment.v1.AttachmentRegistry
import net.fabricmc.fabric.api.attachment.v1.AttachmentSyncPredicate
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.world.entity.Entity

/** Synced, unsaved defense state: 0 is inactive, positive values are Simple Domain strength, -1 is Amplification. */
object BarrierState {
    private val STATE =
        AttachmentRegistry.create<Int>(CursedOath.id("barrier_state")) {
            it.initializer { 0 }.syncWith(ByteBufCodecs.VAR_INT, AttachmentSyncPredicate.all())
        }

    fun initialize() = Unit

    fun get(entity: Entity): Int = entity.getAttached(STATE) ?: 0

    internal fun update(
        entity: Entity,
        value: Int,
    ) {
        if (get(entity) != value) entity.setAttached(STATE, value)
    }
}
